### ISSUE TITLE
Expose the type of service

### ISSUE THREAD

**Issue by gmarty** [fxbox/foxbox#278]
Each service should expose a metadata for its type (like `camera` or `extended colour light`).

For each type of service, we expect certain getters and setters (like an `extended colour light` should be able to be turned on and off, or change the brightness and colour).

cc @cr

**Comment by Yoric** [comment]
For the moment, we do the opposite. That is, rather than exposing the fact that the service is an extended color light, we expose the fact that:
- it can be turned on and off;
- the brightness can be changed;
- the color can be changed.

What would this mechanism improve?

Also, this discussion should probably be in the Taxonomy repo.

**Comment by gmarty** [comment]
If I'm a user it makes more sense to know what type of device I'm interacting with that the capabilities it has.

In the frontend lib, we're going to have user(developer)-friendly classes to interact with the services. For example a service which type is `extended colour light` will have JavaScript methods like:

``` js
service.turnOn();
service.setBrightness('50%');
service.setColour('#f60');
```

It will also make it easy to display an icon based on the type rather than on a list of capabilities.

**Comment by cr** [comment]
Also by having services announce a well-known device type we can employ tests that check for expected setters/getters for each type. Else there's a risk that adapter devs will go completely free-form in their choices, leading to unnecessary incompatibilities.

**Comment by Yoric** [comment]
Fair enough. For the time being, we can put this in the service's property bag.

**Comment by cr** [comment]
Ok, that's what I suggested to use for a start, too.

**Comment by samgiles** [comment]
I wonder if an aspect oriented approach such as traits makes sense?

For example:
I have a light. A light implements Switch, where switch defines an on and
an off state.

A light may also implement Colour. Where an RGBa value can be set or
queried (why an alpha chan? Because we can).

A light may also implement Orientation.  Where we can set some kind of
"look at" vector, or angle of rotation.

A camera, another type may also implement Orientation if it can be
moved/rotated, it may also implement Switch if it can be turned on and off.

You should then be able to query the API for a service to find out which
traits it implements.

We could build (probably generate) quite a nice API for adapter developers
with this method. Bonus: the client API would be incredibly simple to
implement.

**Comment by Yoric** [comment]
@samgiles Are you talking from the point of view of the REST API or the Adapter implementers?

For implementing Adapters, that definitely sounds interesting. For the REST API, I don't really see a difference from what we have atm.

**Comment by samgiles** [comment]
I wrote that pretty sleepy in the plane , but It pretty much is after
looking at the taxonomy code (I haven't had a chance to look at it or play
with it for real yet since I've been networking).

I think my confusion arose when we're talking about defining a fixed type
for a service in this thread, rather than deriving the type from a fixed
set of defined capabilities - I think adapter developers would lose out if
we don't derive from the caps.

**Comment by Yoric** [comment]
See also https://github.com/fxbox/RFC/pull/9 for a change in the Taxonomy architecture that heads a bit in @samgiles' direction.
